javascript/ql/src/semmle/javascript/security/dataflow/UnsafeShellCommandConstructionCustomizations.qll

```diff
@@ -190,6 +190,27 @@ module UnsafeShellCommandConstruction {
     override DataFlow::Node getAlertLocation() { result = this }
   }
 
+  /**
+   * A joined path (`path.{resolve/join}(..)`) that is later executed as a shell command.
+   * Joining a path is similar to string concatenation that automatically inserts slashes.
+   */
+  class JoinedPathEndingInCommandExecutionSink extends Sink {
+    DataFlow::MethodCallNode joinCall;
+    SystemCommandExecution sys;
+
+    JoinedPathEndingInCommandExecutionSink() {
+      this = joinCall.getAnArgument() and
+      joinCall = DataFlow::moduleMember("path", ["resolve", "join"]).getACall() and
+      joinCall = isExecutedAsShellCommand(DataFlow::TypeBackTracker::end(), sys)
+    }
+
+    override string getSinkType() { result = "Uncontrolled path" }
```

Comment:
Minor issue:
I agree with the intent, but I think this string results in a confusing error message.
I fear the user will treat this as a false positive from `js/path-injection` since it is not clear from the message that the problem with the path is that it may be treated as more than just a file path.

Reply:
The full alert message with the current implementation is:  
`"Uncontrolled path based on library input is later used in shell command."`

I agree that the mix of paths and shellscript terminology can be confusing, but I don't have a better idea right now. 

Maybe calling it a `"Path concatenation"` could work better?

Reply:
> Maybe calling it a "Path concatenation" could work better?

Agreed. That also mirrors the "String concatenation" sink type above.